### DISCUSSION TITLE
Windowing with iced?

### DISCUSSION THREAD

**Discussion by 2e0byo** [emacs-ng/emacs-ng#504] [Ideas]
Does anyone have any experience with [iced](https://docs.rs/iced_native/0.9.1/iced_native/index.html) ?  Iced native can apparently expose a `raw_window_handle`, and all the widgets we need are already built in (tooltips, menus).

**Comment by declantsien** [comment]
2e0byo ***@***.***> writes:

> Does anyone have any experience with [iced](https://docs.rs/iced_native/0.9.1/iced_native/index.html) ?  Iced native can apparently expose a `raw_window_handle`, and all the widgets we need are already built in (tooltips, menus).
>
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/emacs-ng/emacs-ng/discussions/504
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>

Hum, I can't picture a Iced use case in Emacs. Emacs draws everything
itself, except the menu. But tao has menu support already.

**Reply by 2e0byo** [reply]
and the tooltips! Iced has a tooltip widget *but* I've no idea how it plays with just treating the window as a surface to draw on.

I was actually looking for some tooltip implementations to have a go at writing one for tao.

However I think iced just draws the tooltip in the viewport without using a new surface at all. [src](https://docs.rs/iced_native/0.9.1/src/iced_native/widget/tooltip.rs.html#171-221).  But I've not read the source thoroughly enough.

**Comment by declantsien** [comment]
2e0byo ***@***.***> writes:

> and the tooltips! Iced has a tooltip widget *but* I've no idea how it plays with just treating the window as a surface to draw on.
>
> I was actually looking for some tooltip implementations to have a go at writing one for tao.
>
> However I think iced just draws the tooltip in the viewport without using a new surface at all. [src](https://docs.rs/iced_native/0.9.1/src/iced_native/widget/tooltip.rs.html#171-221).  But I've not read the source thoroughly enough.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/emacs-ng/emacs-ng/discussions/504#discussioncomment-5189322
> You are receiving this because you commented.
>
> Message ID: ***@***.***>

Emacs has an overlay(not sure whether is called overlay) implantation of
tooltip, which works with tao already. Just tried with `company-mode`.

For native tooptip, there is gtk::Tooltip[0], since tao on Unix is
basically GTK window. Doubt that gonna work with
`--with-wr-gl=glutin/surfman`.
I think for GTK widget to work, we need `--with-wr-gl=gtk3`.


[0](https://gtk-rs.org/gtk3-rs/stable/latest/docs/gtk/struct.Tooltip.html)

**Reply by 2e0byo** [reply]
yes, the overlay tooltips work.  Actually now I think of it there are two kinds of tooltips: company-mode style selection window (which work) and lsp like large tooltips, which currently spawns a real window and then crashes (segfault in gtk, at least on hyprland).  I'll open an issue about that.
